lib/src/log_handler.dart

```diff
@@ -0,0 +1,99 @@
+import 'dart:convert';
+
+import 'package:file/file.dart';
+import 'package:path/path.dart' as p;
+
+import 'constants.dart';
+import 'initializer.dart';
+
+class LogFileStats {
+  /// The oldest timestamp in the log file
+  final DateTime startDateTime;
+
+  /// The latest timestamp in the log file
+  final DateTime endDateTime;
+
+  /// The number of unique session ids found in the log file
+  final int sessionCount;
+
+  /// Contains the data from the [LogHandler.logFileStats]
+  LogFileStats({
+    required this.startDateTime,
+    required this.endDateTime,
+    required this.sessionCount,
+  });
+
+  @override
+  String toString() => jsonEncode(<String, dynamic>{
+        'startDateTime': startDateTime.toString(),
+        'endDateTime': endDateTime.toString(),
+        'sessionCount': sessionCount,
+      });
+}
+
+class LogHandler {
+  final FileSystem fs;
+  final Directory homeDirectory;
+  final File logFile;
+
+  /// This class is responsible for writing to a log
+  /// file that has been initialized by the [Initializer]
+  ///
+  /// It will be treated as an append only log and will be limited
+  /// to have has many data records as specified by [kLogFileLength]
+  LogHandler({
+    required this.fs,
+    required this.homeDirectory,
+  }) : logFile = fs.file(p.join(
+          homeDirectory.path,
+          kDartToolDirectoryName,
+          kLogFileName,
+        ));
+
+  /// Get stats from the persisted log file
+  LogFileStats? logFileStats() {
+    Iterable<Map<String, dynamic>> records =
```

Comment:
Got it. Just want to confirm if I refactored this correctly

OLD
```dart
    // Collection of unique sessions
    final Set<int> sessions = <int>{};
    for (Map<String, dynamic> element in records) {
      sessions.add(element['user_properties']['session_id']['value']);
    }
    final int sessionCount = sessions.length;
```

NEW
```dart
    // Collection of unique sessions
    final Set<int> sessions = <int>{};
    for (Map<String, Object?> element in records) {
      final Map<String, int> prop =
          element['user_properties'] as Map<String, int>;
      final int? value = prop['value'];
      if (value != null) sessions.add(value);
    }
    final int sessionCount = sessions.length;
```

Reply:
This looks right. Can `element['user_properties']['session_id']['value']` ever be null?

Reply:
Well, of course anything is possible. Should it ever be null?

Reply:
As long as the log file hasn't been touched, there should always be a value for `session_id`. The structure of the json is determined by the schema needed by GA and a snippet is shown below

```
{
  'user_properties': {
    'session_id`: {
      'value': 12345678
    }
    ....
  }
}
```

So maybe instead of `int?`, I can make it `final int value = prop['value']!;`? and do away with the conditional check?

Reply:
But now that I think I about it, is it safe to assume that any of the keys exist? Should I perform checks to make sure that `user_properties`, `session_id` and `value` exist? And if they don't just skip that record or return null from the `logFileStats()` method?

Reply:
So with this kind of thing, there's always a balance of:

1. creating signal for ourselves that we have a bug
2. not disrupting what the user was doing

In this case, since this log persistence is not on the critical path of sending analytics events, I think it makes sense to skip the record if it is malformed (once we have this hooked up we'd obviously want to verify it's at least working locally for us).